### PR TITLE
Release --> Master

### PR DESCRIPTION
**Pull request by Jesperobb**
@gerritvanos

### CODE REVIEW

File: code/src/qik_2s12v10.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    const uint8_t qik_2s12v10_c::qik_motor_m0_set_forward = 0x88;
+    const uint8_t qik_2s12v10_c::qik_motor_m0_set_reverse = 0x8A;
+    const uint8_t qik_2s12v10_c::qik_motor_m1_set_forward = 0x8C;
+    const uint8_t qik_2s12v10_c::qik_motor_m1_set_reverse = 0x8E;
+    const uint8_t qik_2s12v10_c::qik_get_config_parameter = 0x83;
+    const uint8_t qik_2s12v10_c::qik_get_error = 0x82;
+
+    qik_2s12v10_c::qik_2s12v10_c(r2d2::uart_ports_c uart_port,
+                                 unsigned int baudRate,
+                                 hwlib::pin_out *_reset_pin)
```
Comment (gerritvanos) [inline code/src/qik_2s12v10.cpp:16]: why is this a pointer and not a reference

File: code/src/qik_2s12v10.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        usart_bus << qik_get_config_parameter << parameter; // send request
+        while (!usart_bus.available()) {
+        }                           // wait for answer
+        return usart_bus.receive(); // return answer
+    }
+    void qik_2s12v10_c::set_configuration_parameter(const uint8_t &parameter,
+                                                    const uint8_t &value) {
+    }
+
+    uint8_t qik_2s12v10_c::get_m0_current() {
```
Comment (gerritvanos) [inline code/src/qik_2s12v10.cpp:100]: is this still wip??

File: code/headers/qik_2s12v10.hpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        static const uint8_t qik_autodetect_baudrate;
+        static const uint8_t qik_request_firmwareversion;
+        static const uint8_t qik_motor_m0_set_forward;
+        static const uint8_t qik_motor_m0_set_reverse;
+        static const uint8_t qik_motor_m1_set_forward;
+        static const uint8_t qik_motor_m1_set_reverse;
+        static const uint8_t qik_get_config_parameter;
+        static const uint8_t qik_get_error;
+
+        hwlib::pin_out *reset_pin;
```
Comment (gerritvanos) [inline code/headers/qik_2s12v10.hpp:33]: same as ohter comment why is this a pointer

File: code/headers/qik_2s12v10.hpp [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        /// Qik2s12v10 constructor.
+        /// @details
+        /// Constructs a Qik2s12v10 with the given parameters.
+        /// @param uart_port The port that the qik_2s12v10_c class will use to
+        /// communicate with the qik 2s12v10 motor controller board.
+        /// @param baudrate The baud rate that the Qik2s12v10 will use in its
+        /// UART TTL serial communication in bps.
+        /// @param _reset_pin Pin that can be used to reset the qik2s12v10.
+        qik_2s12v10_c(r2d2::uart_ports_c uart_port, unsigned int baudrate,
+                      hwlib::pin_out *_reset_pin);
```
Comment (gerritvanos) [inline code/headers/qik_2s12v10.hpp:47]: why the _ infront of reset_pin this is not necessary

File: code/headers/qik_2s12v10.hpp [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+        /// @brief
+        /// Sets the speed of both motors.
+        /// @param _speed The speed of both motors. This value can be between
+        /// -128 and 127, where -128 is full power backwards, 0 is no power and
+        /// 127 is full power forward.
+        void set_speed(const int8_t &_speed) override;
```
Comment (gerritvanos) [inline code/headers/qik_2s12v10.hpp:54]: same underscore comment in all functions

**Comment by Obbesjesper** [comment]
@ThePatman3

File: code/headers/qik_2s12v10.hpp (hunk 5 above)
Comment (gerritvanos) [inline code/headers/qik_2s12v10.hpp:54]: same underscore comments

File: code/src/qik_2s12v10.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
+#include "qik_2s12v10.hpp"
+
+namespace r2d2::moving_platform {
+
+    const uint8_t qik_2s12v10_c::qik_autodetect_baudrate = 0xAA;
```
Comment (gerritvanos) [inline code/src/qik_2s12v10.cpp:5]: why not initialize these in your hpp this is a lot of unnecessary code

File: code/src/qik_2s12v10.cpp (hunk 2 above)
Comment (gerritvanos) [reply code/src/qik_2s12v10.cpp:100]: this is for all functions below

File: code/headers/moving_platform.hpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        /**
+		 * functiuons for testing purpose
+		 * !not te be implemented or used in final product!
+		 */
+        virtual void move(const int8_t &distance) = 0;
+        virtual void move(const int8_t &x, const int8_t &y) = 0;
+
+    public:
+        moving_platform_c(base_comm_c &comm) : base_module_c(comm) {
+            //comm.listen_for_frames();
```
Comment (gerritvanos) [inline code/headers/moving_platform.hpp:54]: why is this commented?

File: code/headers/qik_2s12v10.hpp (hunk 5 above)
Comment (gerritvanos) [reply code/headers/qik_2s12v10.hpp:54]: maybe consider the name change to new_speed this i more clear

File: code/src/qik_2s12v10.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        // reset the qik 2s12v10
+        reset_pin->write(0);
+        hwlib::wait_ms(1);
+        reset_pin->write(1);
+        hwlib::wait_ms(10);
+
+        // setup the uart communication
+        usart_bus << qik_autodetect_baudrate;
+    }
+    void qik_2s12v10_c::set_m0_brake(const uint8_t &brake) {
```
Comment (gerritvanos) [inline code/src/qik_2s12v10.cpp:70]: this also still WIP?

File: code/headers/moving_platform.hpp [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+	 */
+    class moving_platform_c : public base_module_c {
+    protected:
+        /**
+		 * this speed represents a throtle in percentages (%)
+		 * This means that 100 is max and forward, -100 is backwards
+		 * 
+		 * the angle is represented in degrees
+		 */
+        int8_t speed, angle;
```
Comment (itzandroidtab) [inline code/headers/moving_platform.hpp:19]: Is a int8_t enough for the angle? An angle can be between 0 and 360 right?

File: code/headers/moving_platform.hpp [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+		
+		/**
+		 * set the angle to the given value
+		 */
+        virtual void set_steering(const int8_t &steering) = 0;
+
+		/**
+		 * returns the speed value
+		 */
+        inline signed int8_t get_speed() const {
```
Comment (itzandroidtab) [inline code/headers/moving_platform.hpp:34]: Why are these functions inline? The compiler should do this by itself if it wants to optimize. There is no need to force the compiler in a tight corner just to inline the function.

File: code/headers/moving_platform.hpp [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+            return speed;
+        };
+		
+		/**
+		 * returns the angle value
+		 */
+        inline signed int8_t get_steering() const {
```
Comment (itzandroidtab) [inline code/headers/moving_platform.hpp:41]: See other comment above

File: code/headers/moving_platform.hpp [hunk 12]
```diff
@@ -0,0 +1,65 @@
+#pragma once
```
Comment (itzandroidtab) [inline code/headers/moving_platform.hpp:1]: This file is not formatted. There are tabs mixed with 4 spaces

File: code/headers/qik_2s12v10.hpp [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    /// motors. The qik2s12v10 can control the speed and brakes
+    /// of the individual motors. The qik2s12v10 can also set and return its
+    /// configuration parameter, which includes things like Acceleration rate
+    /// and PWM resolution and frequency. Lastly, the qik2s12v10 can check
+    /// for errors and return these.
+    /// For full documentation on the qik2s12v10, see this site:
+    /// https://www.pololu.com/docs/0J29
+    class qik_2s12v10_c : motor_control_c {
+    private:
+        static const uint8_t qik_autodetect_baudrate;
```
Comment (itzandroidtab) [inline code/headers/qik_2s12v10.hpp:24]: Why are you making a static const variable that has no value?

File: code/headers/qik_2s12v10.hpp [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+        /// Sets the brake for both motors.
+        /// @param brake The amount of brake that will be applied to the motors,
+        /// in the range [0,128] where 0 is no brake and 127 is full brake.
+        void set_brakes(const uint8_t &brake);
+
+        /// @brief
+        /// Returns the errors that the qik2s12v10 has detected since this
+        /// function was last used. The meaning of each bit can be found here:
+        /// https://www.pololu.com/docs/0J29/5.c
+        uint8_t get_error();
```
Comment (itzandroidtab) [inline code/headers/qik_2s12v10.hpp:92]: Maybe there should be an enum to compare the error messages to.

File: code/headers/qik_2s12v10.hpp [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+        /// @brief
+        /// Qik2s12v10 constructor.
+        /// @details
+        /// Constructs a Qik2s12v10 with the given parameters.
+        /// @param uart_port The port that the qik_2s12v10_c class will use to
+        /// communicate with the qik 2s12v10 motor controller board.
+        /// @param baudrate The baud rate that the Qik2s12v10 will use in its
+        /// UART TTL serial communication in bps.
+        /// @param _reset_pin Pin that can be used to reset the qik2s12v10.
+        qik_2s12v10_c(r2d2::uart_ports_c uart_port, unsigned int baudrate,
```
Comment (itzandroidtab) [inline code/headers/qik_2s12v10.hpp:46]: Why are you making a copy of the uart_port? This should be as a reference.

File: code/main-qik_2s12v10_c-testing.cpp [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    qik_2s12v10_motorcontroller.init();
+    qik_2s12v10_motorcontroller.get_configuration_parameter(1);
+    hwlib::cout << "Initialization compleet.\n";
+
+    if (test_set_speed) {
+        hwlib::cout << "Testing both motors, full power forward.\n";
+        qik_2s12v10_motorcontroller.set_speed(127);
+        hwlib::wait_ms(2000);
+        hwlib::cout << "Testing both motors, full power backward.\n";
+        qik_2s12v10_motorcontroller.set_speed(-127);
```
Comment (itzandroidtab) [inline code/main-qik_2s12v10_c-testing.cpp:32]: In the documentation you said that -128 was full power. Is it -127 or -128?

File: code/main-qik_2s12v10_c-testing.cpp [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+        hwlib::wait_ms(2000);
+        qik_2s12v10_motorcontroller.set_speed(0);
+        hwlib::cout << "Motor tests compleet.\n";
+        hwlib::wait_ms(500);
+    }
+    if (test_get_configuration_parameter) {
+        hwlib::cout << "Testing the get_configuration_parameter function.\n";
+        hwlib::cout << "Current PWM configuration (should be 0): "
+                    << qik_2s12v10_motorcontroller.get_configuration_parameter(
+                           1)
```
Comment (itzandroidtab) [inline code/main-qik_2s12v10_c-testing.cpp:56]: Why is this on a single line on its own?

File: code/main.cpp [hunk 18]
```diff
@@ -0,0 +1,12 @@
+#include "hwlib.hpp"
```
Comment (itzandroidtab) [inline code/main.cpp:1]: the include should be changed to use ```<>```

File: code/headers/qik_2s12v10.hpp (hunk 13 above)
Comment (itzandroidtab) [reply code/headers/qik_2s12v10.hpp:24]: Okay they are defined somewhere else. You should add some comment's to the code or you should move the definition to here

File: code/src/qik_2s12v10.cpp [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+    }
+    void qik_2s12v10_c::set_brakes(const uint8_t &brake) {
+    }
+
+    uint8_t qik_2s12v10_c::get_error() {
+        while (usart_bus.available() > 0) {
+            usart_bus.receive();
+        }                           // clear the buffer
+        usart_bus << qik_get_error; // send request
+        while (!usart_bus.available()) {
```
Comment (itzandroidtab) [inline code/src/qik_2s12v10.cpp:82]: This infinite loop is kinda dangerous in a RTOS system. In the doxygen there is no documentation that this has an endless loops.

File: code/src/qik_2s12v10.cpp [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+        }                           // wait for answer
+        return usart_bus.receive(); // return answer
+    }
+    uint8_t
+    qik_2s12v10_c::get_configuration_parameter(const uint8_t &parameter) {
+        while (usart_bus.available() > 0) {
+            usart_bus.receive();
+        }                                                   // clear the buffer
+        usart_bus << qik_get_config_parameter << parameter; // send request
+        while (!usart_bus.available()) {
```
Comment (itzandroidtab) [inline code/src/qik_2s12v10.cpp:92]: See comment above

File: code/src/qik_2s12v10.cpp [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+        }                           // wait for answer
+        return usart_bus.receive(); // return answer
+    }
+    void qik_2s12v10_c::set_configuration_parameter(const uint8_t &parameter,
```
Comment (itzandroidtab) [inline code/src/qik_2s12v10.cpp:96]: Why is there an empty function?

File: code/headers/moving_platform.hpp [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+		 * This means that 100 is max and forward, -100 is backwards
+		 * 
+		 * the angle is represented in degrees
+		 */
+        int8_t speed, angle;
+
+		/**
+		 * set the speed to the given value
+		 */
+        virtual void set_speed(const int8_t &speed) = 0;
```
Comment (itzandroidtab) [inline code/headers/moving_platform.hpp:24]: Is a int8_t large enough for the motor speed? Are we losing to much precision on this?

File: code/headers/qik_2s12v10.hpp (hunk 15 above)
Comment (itzandroidtab) [inline code/headers/qik_2s12v10.hpp:46]: Is the baudrate of the qik2s12v10 fixed or is it possible to change it?

File: code/headers/qik_2s12v10.hpp [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+        /// Returns the raw reading from motor M1 that indicates how much
+        /// current flows through the motor at average over the last 5ms. This
+        /// reading is raw, so not converted to milliampere.
+        uint8_t get_m1_current();
+        /// @brief
+        /// This function uses the raw reading from motor M0 to estimate how
+        /// much current flows through the motor in milliampere. Note that the
+        /// value returned by this function can differ from the actual current
+        /// by as much as 20%.
+        unsigned int get_m0_current_milliamps();
```
Comment (itzandroidtab) [inline code/headers/qik_2s12v10.hpp:124]: Is it not better to return a size_t or a uint16_t for the milliamps? Then we have a fixed size for these variables

File: code/main-qik_2s12v10_c-testing.cpp [hunk 24]
```diff
@@ -0,0 +1,71 @@
+/*
```
Comment (itzandroidtab) [inline code/main-qik_2s12v10_c-testing.cpp:1]: This file should be moved to a examples folder in the root of the module.

**Review by itzandroidtab — CHANGES_REQUESTED** [review]

File: code/headers/motor_control.hpp [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+
+	/**
+	 * Class motor_control_c is an abstract class that can be implemented for all different kinds of motors
+	 * This class exists so you can use the motor even without knowing what motor you have been provided with. 
+	 */
+    class motor_control_c {
+    protected:
+		/**
+		 * this speed represents a throtle in percentages (%)
+		 * This means that 100 is max and forward, -100 is backwards
```
Comment (itzandroidtab) [inline code/headers/motor_control.hpp:14]: Why does this value have a different range of the ```set_speed()``` than the qik_2s12v10.hpp?

File: code/headers/motor_control.hpp [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+		 */
+        int8_t speed;
+
+    public:
+		/**
+		 * the speed set must not exceed 100
+		 */
+        virtual void set_speed(const int8_t &_speed) = 0;
```
Comment (itzandroidtab) [inline code/headers/motor_control.hpp:22]: same as the variable

File: code/headers/moving_platform.hpp [hunk 27]
```diff
@@ -0,0 +1,65 @@
+#pragma once
+#include "hwlib.hpp"
```
Comment (itzandroidtab) [inline code/headers/moving_platform.hpp:2]: wrong inlcude ```<>```

File: code/headers/qik_2s12v10.hpp [hunk 28]
```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include "hwlib.hpp"
```
Comment (itzandroidtab) [inline code/headers/qik_2s12v10.hpp:3]: these include's should all changed to includes with ```<>```

**Comment by itzandroidtab** [comment]
So most of the include's are with ```""```. All of them should change to ```<>```.

File: code/src/qik_2s12v10.cpp (hunk 19 above)
Comment (gerritvanos) [reply code/src/qik_2s12v10.cpp:82]: could be solved to add a wait i guess

File: code/headers/moving_platform.hpp (hunk 10 above)
Comment (gerritvanos) [reply code/headers/moving_platform.hpp:34]: also why is it signed int8_t this is signed already

File: code/src/qik_2s12v10.cpp [hunk 29]
```diff
@@ -0,0 +1,113 @@
...
+    const uint8_t qik_2s12v10_c::qik_request_firmwareversion = 0x81;
+    const uint8_t qik_2s12v10_c::qik_motor_m0_set_forward = 0x88;
+    const uint8_t qik_2s12v10_c::qik_motor_m0_set_reverse = 0x8A;
+    const uint8_t qik_2s12v10_c::qik_motor_m1_set_forward = 0x8C;
+    const uint8_t qik_2s12v10_c::qik_motor_m1_set_reverse = 0x8E;
+    const uint8_t qik_2s12v10_c::qik_get_config_parameter = 0x83;
+    const uint8_t qik_2s12v10_c::qik_get_error = 0x82;
+
+    qik_2s12v10_c::qik_2s12v10_c(r2d2::uart_ports_c uart_port,
+                                 unsigned int baudRate,
```
Comment (gerritvanos) [inline code/src/qik_2s12v10.cpp:15]: this variable is camelcase

**Pull request closed without merging**